packages/cw-denom/src/lib.rs

```diff
@@ -0,0 +1,336 @@
+//! This is a simple package for validating cw20 and Cosmos SDK native
+//! denominations. It proves the types, `UncheckedDenom` and
+//! `CheckedDenom`. `UncheckedDenom` may be used in CosmWasm contract
+//! messages and checked via the `into_checked` method.
+//!
+//! To validate native denominations, this package uses the [same
+//! rules](https://github.com/cosmos/cosmos-sdk/blob/7728516abfab950dc7a9120caad4870f1f962df5/types/coin.go#L865-L867)
+//! as the SDK.
+//!
+//! To validate cw20 denominations this package ensures that the
+//! specified address is valid, that the specified address is a
+//! CosmWasm contract, and that the specified address responds
+//! correctly to cw20 `TokenInfo` queries.
+
+#[cfg(test)]
+mod integration_tests;
+
+use std::fmt::{self};
+
+use cosmwasm_std::{
+    to_binary, Addr, BankMsg, Coin, CosmosMsg, CustomQuery, Deps, QuerierWrapper, StdError,
+    StdResult, Uint128, WasmMsg,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum DenomError {
+    #[error(transparent)]
+    Std(#[from] StdError),
+
+    #[error("invalid cw20 - did not respond to `TokenInfo` query: {err}")]
+    InvalidCw20 { err: StdError },
+
+    #[error("invalid native denom. length must be between in [3, 128], got ({len})")]
+    NativeDenomLength { len: usize },
+
+    #[error("expected alphabetic ascii character in native denomination")]
+    NonAlphabeticAscii,
+
+    #[error("invalid character ({c}) in native denom")]
+    InvalidCharacter { c: char },
+}
+
+/// A denom that has been checked to point to a valid asset. This enum
+/// should never be constructed literally and should always be built
+/// by calling `into_checked` on an `UncheckedDenom` instance.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum CheckedDenom {
+    /// A native (bank module) asset.
+    Native(String),
+    /// A cw20 asset.
+    Cw20(Addr),
+}
+
+/// A denom that has not been checked to confirm it points to a valid
+/// asset.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum UncheckedDenom {
+    /// A native (bank module) asset.
+    Native(String),
+    /// A cw20 asset.
+    Cw20(String),
+}
+
```

Comment:
I've added `Asset` type to combine denom with amount:
https://github.com/42-Hackwasm/yield-aggregator/blob/main/packages/cw-denom/src/lib.rs#L59

Maybe we can copy it over here or do in other PR?

Reply:
i actually intentionally did not include the asset part here. i feel that cw20 and native tokens have reasonably distinct interfaces that are difficult to reconcile. for example, how would you expose something like `transfer_from` on a native asset?

larry0x has a good approach with [`cw-asset`](https://github.com/mars-protocol/cw-asset), but IMO i don't want to use it in the contracts because it is possible to call a method and have it panic if the underlying asset type isn't correct. avoiding panics where possible is really important to me.

would love to hear more thoughts here. :) this is not to say that we should not make an asset library, just to say that i'm not sure how it is best done.

Reply:
Asset is just a base object. `cw20` (with `transfer_from`, `send` etc...) is a specialization.

But if it's not needed here, then fair enough to stick with `cw-asset`.